### PR TITLE
`parachain-omni-node` binary

### PR DESCRIPTION
**Pull request by kianenigma**
Related to #62, https://github.com/paritytech/polkadot-sdk/issues/5, https://github.com/paritytech/polkadot-sdk/issues/1984 and possibly more.

This node now should be able to take any runtime that has no consensus related expectations (ie. data being in header digests etc), and run it with its own manual consensus. It should be ideal for a scenario like a light tutorial. 

But moreover, if the approach here is good, it can already be used to build `solo-chain-omni-node` and `parachain-omni-node` etc.

### CODE REVIEW

File: templates/minimal/node/src/command.rs [hunk 1]
```diff
@@ -51,12 +56,34 @@ impl SubstrateCli for Cli {
...
+		let code = std::fs::read(&self.runtime)
+			.map_err(|e| format!("Failed to runtime read {}: {}", &self.runtime, e))?;
+
+		Ok(Box::new(if maybe_path.is_empty() {
+			println!("Using development chain spec, no genesis state set.");
+			let mut properties = Properties::new();
+			properties.insert("tokenDecimals".to_string(), 0.into());
+			properties.insert("tokenSymbol".to_string(), "MINI".into());
+
+			// `.with_genesis_config(Default::default)` won't work, but should.
```
Comment (kianenigma) [inline templates/minimal/node/src/command.rs:72]: @michalkucharczyk this was my main chain-spec related issue so far, I would have expected it to work without this. Or perhaps the `Default::default` should do this for me.

File: templates/minimal/node/src/impl_fake_runtime_apis.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#![allow(unused_variables)]
+
+pub struct FakeRuntime;
+use crate::types::{AccountId, Header, Nonce, OpaqueBlock as Block};
+
+sp_api::impl_runtime_apis! {
```
Comment (kianenigma) [inline templates/minimal/node/src/impl_fake_runtime_apis.rs:6]: We can also manually implement this stuff without this macro, but this macro already does the job easily.

File: templates/minimal/node/src/command.rs (hunk 1 above)
Comment (michalkucharczyk) [reply templates/minimal/node/src/command.rs:72]: Actually, `.with_genesis_config_patch(Default::default)` should do the job. (which means empty patch for default config).

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5596506

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5596541

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: check-tracing
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5596542

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5621470

**Comment by bkchr** [comment]
> parachain-omni-node

`polkadot-parachains` should be this binary. No need to have another special binary for this IMO.

**Comment by kianenigma** [comment]
> > parachain-omni-node
> 
> `polkadot-parachains` should be this binary. No need to have another special binary for this IMO.

Agreed, I didn't know it is already a node capable of this.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/parity-tech-update-for-march/7226/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/rfc-should-we-launch-a-thousand-cores-program/7604/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/paritytech-update-for-april/7646/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/stabilizing-polkadot/7175/20

File: omni-nodes/common/src/fake_runtime.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	pub(crate) type Nonce = u32;
+	pub(crate) type AuraId = sp_consensus_aura::sr25519::AuthorityId;
+	pub(crate) type BlockNumber = u32;
+	pub(crate) type Balance = u128;
+	pub(crate) type Weight = sp_weights::Weight;
+	pub(crate) type RuntimeCall = ();
+}
+use whatever::*;
+
+sp_api::impl_runtime_apis! {
```
Comment (kianenigma) [inline omni-nodes/common/src/fake_runtime.rs:30]: This should ideally move to `cumulus_service` now. I think one set of fake runtime impls should be enough.

File: omni-nodes/parachain/src/chain_spec.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		Extensions {
+			relay_chain: "rococo-local".into(),
+			// You MUST set this to the correct network!
+			para_id: 1000,
+		},
+	)
+	.with_name("Development")
+	.with_id("dev")
+	.with_chain_type(ChainType::Development)
+	.with_genesis_config_patch(testnet_genesis(
```
Comment (michalkucharczyk) [inline omni-nodes/parachain/src/chain_spec.rs:82]: Don't know if this is final or temporary proposal. I'd suggest fetching `"testnet"` genesis from runtime here as _omni-node_ shall make no assumptions about runtime.

If _omni-node_ is to provide two "hardcoded" chainspecs (`"development", "local_testnet"`) then we could require runtime's author to provided corresponding presets. If runtime does not provide those presets we could fail.

Comment (michalkucharczyk) [reply omni-nodes/parachain/src/chain_spec.rs:82]: oh, it is not actually used...

**Comment by kianenigma** [comment]
> What is blocking this (assuming it is blocked)?
> 
> Upgrade the client to the new polkadot-sdk is always a big overhead and error-prune process and for no real benefits. Omni node is the solution to allow parachain teams only focus on the runtime. This is a super critical work and I would like to see it released ASAP.

There should be no blocker for syncing normal parachains as far as I can say. I have synced hydra, interlay and frequncy at https://github.com/paritytech/polkadot-sdk/pull/3597/commits/40fbf47f4c28cc7ea8520815d9b8df59cc3597a0 and it all works. Although, as additional tests, we should also test: 

1. If the DApp of the corresponding parachain can work with the omni-node (missing RPCs)
2. If the node can collate blocks (important, else this is not useful)

The code in this commit already provides some soft degree of configurability through a builder pattern as well. This can either remain a builder or become CLI flags. 

https://github.com/paritytech/polkadot-sdk/blob/40fbf47f4c28cc7ea8520815d9b8df59cc3597a0/omni-nodes/parachain/src/main.rs#L14-L20

**Comment by xlc** [comment]
let’s get a simple version out that only supports the most common scenario and then we can iteratively make it compatible with more chains

**Comment by kianenigma** [comment]
https://github.com/paritytech/polkadot-sdk/pull/4369/ is a pretty similar PR to what is being done here, and can be reviewed and merged as a first step.

**Comment by kianenigma** [comment]
This branch will likely be closed in favor of s slimmer version brought to you by @serban300. I will leave open for visibility for now, but feel free to close when a replacemebt comes by.

**Pull request closed without merging**